images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/modules/i2c_drivers/aht20_driver.py

```diff
@@ -0,0 +1,91 @@
+"""
+[module]
+name=aht20
+enabled=true
+[end]
+"""
+try:
+    from x1plus.services.x1plusd.modules.expansion.i2c import register_driver
+except ImportError:
+    from ..expansion.i2c import register_driver
```

Comment:
ah, yeah, this is the bit I was worried about :) this means that there needs to be a dependency problem of order in how drivers are loaded.  specifically, this does not necessarily load the i2c that came from a module, but from this on-disk one.  so maybe the expansion core drivers really need to be built-in for now, and the i2c drivers should be loadable?

Reply:
Is there much risk with this approach though, if it prioritizes one from the import path and as a backup, uses on-disk relative pathing? If it's not loaded into the system modules yet, it uses the on-disk one up until the on-disk one is imported itself into system-modules, bringing everything with it.

The benefit is for when a user has a custom module on the sd-card - it should always end up using the module path first as custom sd-card modules are all loaded *after* they get added to the system modules.

Then to template, these existing modules in core directories have both import options to get around dependency issues as you mention - which only exist for this decorator type function, due to when and where the modules are actually loaded, and for some, started. I was testing this with completely wiped pycache's and it seems to not matter and all still get added to `i2c` or `ledstrip` or `actions`, despite forcing a dependency order issue. As it gets imported from the disk during x1plusd's `__init__` anyways, it seems to be fully functional and brings everything with it.

While not done here, this similar pattern could also be used for custom sd-card modules that have non-module dependencies in other files. It could hope it's on the path, fail, then load local.

Reply:
Hm, I believe that it worked, but I'm also not sure why it *must* work.  I think my concern is this: if there's one in x1plus in the rootfs, and one on the SD card, then which one you get might not be consistent between modules (i.e., if it was an in-rootfs module or if it was an on-SD-card module) -- otherwise, that `except ImportError:` wouldn't be necessary ever, right?

And, beyond that, even if you do consistently get one loaded from the SD card, it also means that you can end up downgrading something that x1plusd internally depends on across x1plus version upgrades.

I think I am going to haul the i2c and ledstrip drivers back into the core for now.  But we should definitely revisit this in the future.